### PR TITLE
IGNITE-14255: Added the ability to flush all performance statistics buffers to the current file and start recording to the next file.

### PR DESCRIPTION
**Pull request by Sega76**
…uffers to the current file and start recording to the next file.

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [x] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/control-utility/src/test/java/org/apache/ignite/util/PerformanceStatisticsCommandTest.java [hunk 1]
```diff
@@ -73,6 +80,16 @@ public void testCommands() throws Exception {
         assertEquals(EXIT_CODE_OK, res);
         assertEquals(STATUS_ENABLED, lastOperationResult);
 
+        res = execute(PERFORMANCE_STATISTICS.text(), ROTATE.toString());
+
+        assertEquals(EXIT_CODE_OK, res);
```
[the remark this answers is not in the document]
Comment (Sega76) [reply modules/control-utility/src/test/java/org/apache/ignite/util/PerformanceStatisticsCommandTest.java:91]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 2]
```diff
@@ -278,6 +297,30 @@ private void stopWriter() {
...
+                FilePerformanceStatisticsWriter newWriter = new FilePerformanceStatisticsWriter(ctx);
+
+                newWriter.start();
+
+                FilePerformanceStatisticsWriter oldWriter = writer;
+
+                writer = newWriter;
+
+                oldWriter.stop();
+            }
```
[the remark this answers is not in the document]
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:318]: done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 3]
```diff
@@ -214,6 +216,23 @@ public void stopCollectStatistics() throws IgniteCheckedException {
...
+    public void rotateCollectStatistics() throws IgniteCheckedException {
+        A.notNull(metastorage, "Metastorage not ready. Node not started?");
+
+        if (ctx.isStopping())
+            throw new NodeStoppingException("Operation has been cancelled (node is stopping)");
+
+        if (!enabled())
+            throw new IgniteCheckedException("Performance statistics collection not started.");
+
+        ctx.task().execute(PerformanceStatisticsRotateTask.class.getName(),new VisorTaskArgument());
```
[the remark this answers is not in the document]
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java]: done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 4]
```diff
@@ -278,6 +307,31 @@ private void stopWriter() {
         log.info("Performance statistics writer stopped.");
     }
 
+    /** Rotate performance statistics writer. */
+    public void rotateWriter() {
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:311]: Can be private

File: modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsRotateFileTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+        startClientGrid(NODES_CNT);
+    }
+
+    /** @throws Exception If failed. */
+    @Test
+    public void testRotateFile() throws Exception {
+        int cnt = 5;
+
+        assertThrows(null, AbstractPerformanceStatisticsTest::rotateCollectStatistics, IgniteException.class,
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsRotateFileTest.java:88]: `log` can be passed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 6]
```diff
@@ -214,6 +220,29 @@ public void stopCollectStatistics() throws IgniteCheckedException {
         metastorage.write(PERF_STAT_KEY, false);
     }
 
+    /**
+     * Rotate collecting performance statistics.
+     *
+     * @throws IgniteCheckedException If stopping failed.
+     */
+    public IgniteInternalFuture<Void> rotateCollectStatistics() throws IgniteCheckedException {
+        A.notNull(metastorage, "Metastorage not ready. Node not started?");
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:229]: This is unnecessary

Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java (hunk 4 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsRotateFileTest.java (hunk 5 above)
Comment (Sega76) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsRotateFileTest.java]: fixed

**Review by NSAmelchev — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsReader.java [hunk 7]
```diff
@@ -69,15 +69,15 @@
...
-    private static final String UUID_STR_PATTERN =
+    static final String UUID_STR_PATTERN =
         "[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}";
 
     /** File name pattern. */
-    private static final Pattern FILE_PATTERN = Pattern.compile("^node-(" + UUID_STR_PATTERN + ")(-\\d+)?.prf$");
+    static final Pattern FILE_PATTERN = Pattern.compile("^node-(" + UUID_STR_PATTERN + ")(-(\\d+))?.prf$");
+
+    /** Default file read buffer size. */
+    private static final int DFLT_READ_BUFFER_SIZE = (int)(8 * U.MB);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsReader.java:80]: Why this change? Please, revert moving this constant from 72 to 80 line.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 8]
```diff
@@ -214,6 +220,27 @@ public void stopCollectStatistics() throws IgniteCheckedException {
         metastorage.write(PERF_STAT_KEY, false);
     }
 
+    /**
+     * Rotate collecting performance statistics.
+     *
+     * @throws IgniteCheckedException If rotating failed.
+     */
+    public IgniteInternalFuture<Void> rotateCollectStatistics() throws IgniteCheckedException {
+        if (!allNodesSupports(ctx.discovery().allNodes(), IgniteFeatures.PERFORMANCE_STATISTICS_ROTATE))
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:229]: This check is redundant. Let's remove it.

Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java]: done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsReader.java (hunk 7 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsReader.java]: done

File: modules/core/src/main/java/org/apache/ignite/internal/util/distributed/EmptySerializable.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package org.apache.ignite.internal.util.distributed;
+
+import java.io.Serializable;
+
+/**
+ * Empty serializable.
+ */
+public class EmptySerializable implements Serializable {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/util/distributed/EmptySerializable.java:25]: Why do we need this file?

**Review by NSAmelchev — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/util/distributed/EmptySerializable.java (hunk 9 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/util/distributed/EmptySerializable.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsReader.java [hunk 10]
```diff
@@ -77,7 +77,7 @@
         "[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}";
 
     /** File name pattern. */
-    private static final Pattern FILE_PATTERN = Pattern.compile("^node-(" + UUID_STR_PATTERN + ")(-\\d+)?.prf$");
+    static final Pattern FILE_PATTERN = Pattern.compile("^node-(" + UUID_STR_PATTERN + ")(-(\\d+))?.prf$");
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsReader.java:80]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 11]
```diff
@@ -214,6 +234,33 @@ public void stopCollectStatistics() throws IgniteCheckedException {
         metastorage.write(PERF_STAT_KEY, false);
     }
 
+    /**
+     * Rotate collecting performance statistics.
+     *
+     * @throws IgniteCheckedException If rotating failed.
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:240]: typo: rotating -> rotation.

File: modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java [hunk 12]
```diff
@@ -180,7 +182,9 @@ public DistributedProcess(
                 return;
             }
 
-            finish.apply(p.id,msg.result(), msg.error());
+            finish.apply(p.id, msg.result(), msg.error());
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java:185]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 13]
```diff
@@ -214,6 +234,33 @@ public void stopCollectStatistics() throws IgniteCheckedException {
...
+     */
+    public IgniteInternalFuture<?> rotateCollectStatistics() throws IgniteCheckedException {
+        if (ctx.isStopping())
+            throw new NodeStoppingException("Operation has been cancelled (node is stopping)");
+
+        if (!enabled())
+            throw new IgniteCheckedException("Performance statistics collection not started.");
+
+        return rotateProc.start(UUID.randomUUID(), null).chain(
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:249]: What will happens if rotation process already running?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 14]
```diff
@@ -278,6 +325,32 @@ private void stopWriter() {
...
+
+                    FilePerformanceStatisticsWriter oldWriter = writer;
+
+                    writer = newWriter;
+
+                    oldWriter.stop();
+                }
+            }
+
+            log.info("Performance statistics writer rotated.");
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:345]: Let's log a file that can be used for analysis here.

`Performance statistics writer rotated[writtenFile=001.prf]`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java (hunk 13 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:249]: happens twice

File: modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java (hunk 12 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java (hunk 11 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java]: done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsReader.java (hunk 10 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsReader.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java (hunk 14 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java]: done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 15]
```diff
@@ -278,6 +325,35 @@ private void stopWriter() {
...
+
+                    oldWriter.stop();
+                }
+            }
+
+            if (oldWriter != null)
+                log.info("Performance statistics writer rotated[writtenFile=" + oldWriter.getFile() + "].");
+        }
+        catch (IgniteCheckedException | IOException e) {
+            log.error("Failed to rotate performance statistics writer.", e);
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:351]: Why do we need to log exception here? 
Is it not logged elsewhere?

File: modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java [hunk 16]
```diff
@@ -242,13 +246,20 @@ else if (F.eq(ctx.localNodeId(), p.crdId)) {
      *
      * @param id Process id.
      * @param req Initial request.
+     * @return Future on the result of the finished process.
      */
-    public void start(UUID id, I req) {
+    public IgniteInternalFuture<T2<Map<UUID, R>, Map<UUID, Exception>>> start(UUID id, @Nullable I req) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java:251]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java [hunk 17]
```diff
@@ -188,6 +191,11 @@ public synchronized void stop() {
         started = false;
     }
 
+    /** @return Performance statistics file. */
+    public File getFile() {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java:195]: This method can be package private.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/AbstractPerformanceStatisticsTest.java [hunk 18]
```diff
@@ -42,6 +47,10 @@
     /** */
     public static final long TIMEOUT = 30_000;
 
+    /** File performance statistics pattern {@link FilePerformanceStatisticsReader}. */
+    static final Pattern FILE_PATTERN = Pattern.compile(
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/AbstractPerformanceStatisticsTest.java:51]: Why do we need to copy-paste regexp here?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java (hunk 15 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:349]: we will get an error when getting the result.
I find this information useful here

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java (hunk 17 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/AbstractPerformanceStatisticsTest.java (hunk 18 above)
Comment (Sega76) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/AbstractPerformanceStatisticsTest.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java (hunk 16 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 19]
```diff
@@ -291,4 +365,25 @@ private void write(Consumer<FilePerformanceStatisticsWriter> c) {
         /** This method is called whenever the performance statistics collecting is started. */
         public void onStarted();
     }
+
+    /** Rotate performance statistics future. */
+    protected static class RotateFuture extends GridFutureAdapter<Serializable> {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:370]: Why do we need this new class?

**Review by nizhikov — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java [hunk 20]
```diff
@@ -278,6 +307,29 @@ private void stopWriter() {
...
+            newWriter.start();
+
+            oldWriter = writer;
+
+            writer = newWriter;
+
+            oldWriter.stop();
+        }
+
+        if (oldWriter != null)
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/PerformanceStatisticsProcessor.java:329]: Add `log.isInfoEnabled()`, please
